src/Microsoft.AspNetCore.Rewrite/Internal/IISUrlRewrite/ServerVariables.cs

```diff
@@ -9,16 +9,21 @@ namespace Microsoft.AspNetCore.Rewrite.Internal.IISUrlRewrite
 {
     public static class ServerVariables
     {
+        public const string RequestHeaderPrefix = "HTTP_";
+        public const string ResponseHeaderPrefix = "RESPONSE_";
+
         /// <summary>
-        /// Returns the matching <see cref="PatternSegment"/> for the given <paramref name="serverVariable"/>
+        /// Returns a <see cref="ServerVariable"/>
         /// </summary>
         /// <param name="serverVariable">The server variable</param>
         /// <param name="context">The parser context which is utilized when an exception is thrown</param>
         /// <param name="global">Indicates if the rule being parsed is a global rule</param>
         /// <exception cref="FormatException">Thrown when the server variable is unknown</exception>
-        /// <returns>The matching <see cref="PatternSegment"/></returns>
-        public static PatternSegment FindServerVariable(string serverVariable, ParserContext context, bool global)
+        /// <returns>A <see cref="ServerVariable"/></returns>
+        public static ServerVariable FindServerVariable(string serverVariable, ParserContext context, bool global)
```

Comment:
I think the problem here is that we're conflating two independent scenarios. Server Variables can be read from or assigned to. Previously this code only supported reading. Now you're trying to add assignment, but you've only implemented two of the variable types, request and response headers.

I think this needs to be split up. Revert the changes to FindServerVariable and create a new method FindServerVariableSetter. FindServerVariableSetter looks up the variables and returns a new kind set of types that mutate the request/response when evaluated. E.g. you need to create a type that's the opposite of HeaderSegment, a HeaderSetter, and a UrlSetter, SchemeSetter, etc.. Then when you loop through these Setters in ApplyRule, you don't need a switch statement, you just call Apply on each one.